### ISSUE TITLE
Using hsmm.backward_estimation() as starting point for LOOCV might be biased

### ISSUE THREAD

**Issue by JoKra1** [GWeindel/hsmm_mvpy#34]
Possible problem:
The estimates from hsmm.backward_estimation() are based on the data from all subjects and thus contain information about the data that is omitted purposefully during LOOCV (i.e., the held-out data from a single subject). Since the estimate recovered by the EM, also used during the LOOCV, depends on the starting value the estimate might also end up being biased towards the solution that worked for all subjects.

Proposed solution:
Move the backward estimation into the LOOCV procedure and calculate a backward estimate for each fold (i.e. the remaining data after excluding one subject). Then use those for the remainder of the LOOCV procedure.

This does appear to affect the results from tutorial 3:

Here the original approach:
![loocv using original approach](https://user-images.githubusercontent.com/52180639/202190945-e4141c74-e8b6-435a-ac4f-c57fcb317946.png)

And here the proposed alternative:
![loocv using proposed solution](https://user-images.githubusercontent.com/52180639/202190935-c235f391-db05-4840-be93-bb1a407a9a71.png)

**Comment by JoKra1** [comment]
See [c5aa379](https://github.com/JoKra1/hsmm_mvpy/commit/c5aa379d931c66d39064deaa35e4f36e1a8ca12c) for possible implementation of proposed solution.

**Comment by GWeindel** [comment]
This does make a lot of sense and should be the default behavior of LOOCV (if not the only one). Additionally the new LOOCV profile of Tutorial 3 looks more credible as higher number of bumps get more punished.

I see two solutions for the problem:
1. Use your implementation and rely on the backward estimation
2. Optimize the starting points search independently of other solutions 

Solution 1. has one potentially big downside is that the computation time will take much longer even more given that the ```__loocv_backwards``` function doesn't benefit from the parallelization. I have no clue how to use paralellization efficiently in that case, my guess is that we cannot distribute everything so we have to choose whether to parallelize over subjects or over n_bumps. I guess it is still better to parallelize over subjects as you are doing it. In the current state it would be useful to know whether computation time is still in an acceptable range for usual applications. 

Solution 2. Would be faster but might not be as good to find the optimal set of starting points and would be hardly tractable so I guess it is not advised

Can you test for time difference between both methods?

**Comment by JoKra1** [comment]
Considering efficiency:

In the proposed solution the increase in complexity should be approximately linear in the number of subjects. In the current implementation we have to call ``backward_estimation()`` once and benefit from the parallelization over bumps. Then we call ``loocv_mp()`` and benefit from the parallelization over subjects. In the proposed solution we essentially call ``backward_estimation()`` not once but as many times as we have subjects, but in every call we still benefit from the parallelization over bumps. Since these calls to ``backward_estimation()`` via ``__loocv_backwards()`` still happen before the original parallelized loop in ``loocv_mp()``, the actual LOOCV procedure should remain unaffected and should thus take the same time.

In tutorial 3 there are only 4 subjects but up to six bumps, so I think here this is efficient. However, with more subjects it might actually become more efficient to parallelize over subjects and then essentially call backward_estimation(cpus=1) for every subject (but that is currently not supported I think?).

In that case we should maybe replace: 

```
if bests is None:

        for p in participants:
            if verbose:
                print(f'Estimating LOOCV backwards solutions for participant {p}')

            best_est_p = __loocv_backwards(unstacked_data, p, init.sf, init.bump_width, cpus, max_starting_points)
            bests_est.append(best_est_p)
```

with:

```
if bests is None:
        
        with multiprocessing.Pool(processes=cpus) as pool:
            bests_est = pool.starmap(__loocv_backwards, 
                zip(itertools.repeat(unstacked_data), participants,
                    itertools.repeat(init.sf), itertools.repeat(init.bump_width), 
                    itertools.repeat(1), # Make sure that cpus is 1 for every call to __loocv_backwards()
                    itertools.repeat(max_starting_points)))
```

Here the time data:

In tutorial 3 the call to backward_estimation() for all data reports a wall time of 14.3 seconds on my machine. The loocv_mp() call using these estimates as starting values reports 14.2 seconds.

Calling loocv_mp() with bests=None to obtain backward estimates per fold takes 1 min and 8 secs - which would indeed suggest that the increase is approximately linear in the number of subjects.

I do like the idea of maybe finding an alternative way to optimize the starting points. Maybe backward_estimation() can be adjusted to not drop **every** other bump to recover the next bump-1 solution? I don't have a good idea yet how this selection could be made smarter but depending on the number of bumps which can be ignored this might speed things up. That would reflect a mixture of options 1 & 2. :-)

**Comment by jelmerborst** [comment]
One thing I was just thinking about: can't we just estimate the max nr of bumps, and then for estimations with lower nrs of bumps pick the top X most likely ones?

**Comment by GWeindel** [comment]
@JoKra1 I don't understand your last suggestion

@jelmerborst It is easy to do and might be good to inspect other possible solution but I can't see yet how it addresses the problem raised by Joshua

**Comment by GWeindel** [comment]
Also as a general comment on the problem: With only 4 participants the weight of each on the starting point is very big so LOOCV will differ highly between the two methods, with larger groups it shouldn't be so much of a trouble although I think that larger number of bumps will be more penalized which is a good thing as this is expected when overfitting

**Comment by GWeindel** [comment]
I am working on a new version to find all possible numbers of bumps that also happens to work ~ 10 times faster than ```backward_estimation``` but it will imply slightly rewriting the ```LOOCV``` function so I keep you posted

**Comment by jelmerborst** [comment]
This is fixed with #102
